### PR TITLE
T-404 Don't query asset.meta.shouldWrap in packager, create fake init instead

### PR DESCRIPTION
**Pull request by mischnic**
# ↪️ Pull Request

Fixes the flaky CI tests
Fixup for #4266

In the JSPackager (`concat.js`) we propagate `meta.shouldWrap` of dependencies and assets down the asset graph (recursively). But that means that reading that flag in a packager for a asset that is part of a different bundle is not possible (because packagers run in parallel).

Instead, create a "fake" init function for referenced assets, so that an importing asset doesn't need to care if the importee is actually wrapped.

## 💻 Examples

```js
...

// real init, generated during concatentation
export function $c88483ac41b2a512779ea3eb3fdae11d$init() {
  if (!$c88483ac41b2a512779ea3eb3fdae11d$executed) {
    $c88483ac41b2a512779ea3eb3fdae11d$executed = true;
    $c88483ac41b2a512779ea3eb3fdae11d$exec();
  }

  return $c88483ac41b2a512779ea3eb3fdae11d$exports;
}

console.log("a", $e268b5364d61efe9eddbb634d90bc$export$default);
import("./" + "b.9cbbbfe7.js");
// fake init, inserted before generating exports
export function $e268b5364d61efe9eddbb634d90bc$init() {
  return $e268b5364d61efe9eddbb634d90bc$exports;
}
```

### CODE REVIEW

File: packages/shared/scope-hoisting/src/link.js [hunk 1]
```diff
@@ -418,6 +424,7 @@ export function link({
           }
 
           if (node) {
+            console.log(2, node);
```
Comment (wbinnssmith) [inline packages/shared/scope-hoisting/src/link.js]: debug 😄

File: packages/shared/scope-hoisting/src/link.js [hunk 2]
```diff
@@ -374,10 +376,14 @@ export function link({
               mod,
               'js',
             );
-            if (!isUnusedValue(path) || isReferenced) {
-              let name = assertString(mod.meta.exportsIdentifier);
+            let used = !isUnusedValue(path);
+
```
Comment (wbinnssmith) [inline packages/shared/scope-hoisting/src/link.js]: nit: maybe a newline above `let used` rather than in between these?

Comment (mischnic) [reply packages/shared/scope-hoisting/src/link.js]: `isReferenced` and `isValueUsed` are closely related, so I wanted to keep them together, but I moved them closer to the if condition.

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>packages/benchmarks/three 🚨</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | **FAILED** | -0.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

<details><summary>packages/benchmarks/kitchen-sink 🚨</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | **FAILED** | -0.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

<details><summary>packages/benchmarks/react-hn 🚨</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | **FAILED** | -0.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

<details><summary>packages/benchmarks/ak-editor 🚨</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | **FAILED** | -0.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/260827801206129153)**

**Comment by wbinnssmith** [comment]
Let's land this to unblock CI and other PRs, and revert it in #4375.

**Pull request merged**